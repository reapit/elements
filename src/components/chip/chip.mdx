import { Meta, Story, Canvas } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as ChipStories from './chip.stories'

<Meta of={ChipStories} />

# Chip

Similar to the chips used on the `MultiSelect` component, the underlying element is a HTML checkbox, allowing for them to be used in a form and with form libraries.

They are used to toggle filters and the state of pages on click or on form submit.

## Basic Usage

<Canvas of={ChipStories.BasicUsage} />

<RenderHtmlMarkup of={ChipStories.BasicUsage} />

## Chip Group

A group of Chips has it's own wrapper to display responsively in a CSS grid.

<Canvas of={ChipStories.ChipGroupUsage} />

<RenderHtmlMarkup of={ChipStories.ChipGroupUsage} />
