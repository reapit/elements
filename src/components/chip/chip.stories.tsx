import { Chip, ChipGroup } from '.'

export default {
  title: 'Chip',
  component: Chip,
}

export const BasicUsage = {
  render: ({}) => (
    <ChipGroup>
      <Chip>Content</Chip>
    </ChipGroup>
  ),
}

export const ChipGroupUsage = {
  render: ({}) => (
    <ChipGroup>
      <Chip>Content</Chip>
      <Chip>Content</Chip>
      <Chip>Content</Chip>
    </ChipGroup>
  ),
}
