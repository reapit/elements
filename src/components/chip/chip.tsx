import { FC, HTMLAttributes, InputHTMLAttributes } from 'react'
import { cx } from '@linaria/core'
import { ElChipCheckbox, ElChipLabel, ElChipGroup, ElChipGroupInner } from './__styles__'
import { useId } from '../../storybook/random-id'

export interface ChipProps extends InputHTMLAttributes<HTMLInputElement> {}

export const Chip: FC<ChipProps> = ({ children, className, id, ...rest }) => {
  const chipId = useId(id)
  return (
    <>
      <ElChipCheckbox id={id ?? chipId} type="checkbox" {...rest} />
      <ElChipLabel htmlFor={id ?? chipId} className={cx(className)}>
        {children}
      </ElChipLabel>
    </>
  )
}

export const ChipGroup: FC<HTMLAttributes<HTMLDivElement>> = ({ children, className, ...rest }) => (
  <ElChipGroup className={cx(className)} {...rest}>
    <ElChipGroupInner>{children}</ElChipGroupInner>
  </ElChipGroup>
)
