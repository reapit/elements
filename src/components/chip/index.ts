export * from './styles'
export * from './chip'
