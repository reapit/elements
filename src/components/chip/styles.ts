import { styled } from '@linaria/react'
import { Icon } from '../icon'
import type { CSSProperties } from 'react'

interface ElChipCSSProperties extends CSSProperties {
  /** Used to determine the maximum width of the chip because the browser does not support
   * [CSS' advanced attr() syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/attr) */
  '--chip-max-width'?: `var(--size-${string})`
}

interface ElChipProps {
  'data-variant': 'filter' | 'selection'
  style?: ElChipCSSProperties
}

export const ElChip = styled.button<ElChipProps>`
  align-items: center;
  border: none;
  border-radius: var(--comp-chip-border-radius);
  color: var(--comp-chip-colour-text-active);
  cursor: pointer;
  display: grid;
  gap: var(--spacing-2);
  grid-template-columns: auto min-content;
  height: min-content;
  max-width: var(--chip-max-width, auto);
  padding-block: var(--spacing-1);
  padding-inline: var(--spacing-4) var(--spacing-2);
  width: fit-content;

  &[aria-disabled='true'] {
    cursor: not-allowed;
    color: var(--comp-chip-colour-text-disabled);
  }

  &[data-variant='filter'] {
    background: var(--comp-chip-colour-fill-filter-default);

    &:hover {
      background: var(--comp-chip-colour-fill-filter-hover);
    }

    &[aria-disabled='true'],
    &[aria-disabled='true']:hover {
      background: var(--comp-chip-colour-fill-filter-disabled);
    }
  }

  &[data-variant='selection'] {
    background: var(--comp-chip-colour-fill-selection-default);

    &:hover {
      background: var(--comp-chip-colour-fill-selection-hover);
    }

    &[aria-disabled='true'],
    &[aria-disabled='true']:hover {
      background: var(--comp-chip-colour-fill-selection-disabled);
    }
  }

  &:focus-visible {
    box-shadow:
      0px 0px 0px var(--size-px) var(--white),
      0px 0px 0px var(--size-1) var(--purple-300);
    outline: none;
  }
`

interface ElChipLabelProps {
  'data-will-truncate'?: boolean
}

export const ElChipLabel = styled.span<ElChipLabelProps>`
  color: var(--text-primary);

  /* Allows long words to be broken and wrapped onto the next line. */
  overflow-wrap: anywhere;

  /* text-sm/Regular */
  font-family: var(--font-family);
  font-size: var(--font-size-sm);
  font-style: normal;
  font-weight: 400;
  letter-spacing: var(--letter-spacing-sm);
  line-height: var(--line-height-sm);
  text-align: left;

  &[data-will-truncate='true'] {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  [aria-disabled='true'] & {
    color: var(--text-placeholder);
  }
`

export const ElChipClearIcon = styled(Icon)`
  /* NOTE: We only use !important here because the current Icon component
   * does not allow consumer-supplied styles to have a higher specificity */
  color: var(--comp-chip-colour-icon-active) !important;
  font-size: 1rem;
  height: var(--icon_size-s) !important;
  width: var(--icon_size-s) !important;

  [aria-disabled='true'] & {
    color: var(--comp-chip-colour-icon-disabled) !important;
  }
`
