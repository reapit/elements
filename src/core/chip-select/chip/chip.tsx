import {
  ElChipSelectChip,
  ElChipSelectChipIconContainer,
  ElChipSelectChipInput,
  ElChipSelectChipLabelText,
} from './styles'
import { forwardRef, useCallback } from 'react'
import { maybeDeselectOtherOptions } from './maybe-deselect-other-options'

import type { ChangeEventHandler, MouseEventHandler, InputHTMLAttributes, ReactNode } from 'react'

// We omit a few attributes from the base input element:
// - onClick, because we want the onClick to be handled by the label, not the input
// - size, because we use this for our own sizing
// - type, because chip select options are always checkboxes
type AttributesToOmit = 'onClick' | 'size' | 'type'

export namespace ChipSelectChip {
  export interface Props extends Omit<InputHTMLAttributes<HTMLInputElement>, AttributesToOmit> {
    /** The accessible name of the chip. Should be considered mandatory when no visual label is provided. */
    'aria-label'?: string
    /** The visual label of the chip. */
    children?: ReactNode
    /** Whether the chip is disabled. Disabled chips, even if checked, will not be submitted. */
    disabled?: boolean
    /**
     * The icon of the chip. All chips in the same chip select should either have an icon or not.
     * If there is no visual label provided via `children`, the icon should be considered mandatory.
     */
    icon?: ReactNode
    /**
     * Whether the chip is exclusive. That is, when it's selected, all other related chips will be
     * deselected. Exclusive and non-exclusive chips should not be mixed in the same chip select.
     */
    isExclusive?: boolean
    /** The maximum width of the chip. If not provided, the chip will be as wide as its content. */
    maxWidth?: `--size-${string}`
    /** Name of the form control. Submitted with the form as part of a name/value pair. */
    name?: string
    /** Callback called when the chip's checked state changes. */
    onChange?: ChangeEventHandler<HTMLInputElement>
    /** Callback called when the chip is clicked. */
    onClick?: MouseEventHandler<HTMLLabelElement>
    /** Whether the label of the chip should be truncated if it is too long */
    overflow?: 'truncate'
    /** Whether the chip is read-only. */
    readOnly?: boolean
    /** The size of the chip. All chips in the same chip select should have the same size. */
    size: 'small' | 'medium' | 'large'
    /** The value of the form control. */
    value: InputHTMLAttributes<HTMLInputElement>['value']
  }
}

/** @deprecated Use ChipSelectChip.Props instead */
export type ChipSelectChipProps = ChipSelectChip.Props

/**
 * An option for a `ChipSelect`. It is a styled native checkbox input, so it's checked state can be
 * controlled (or uncontrolled) like any other native input. Typically used via `ChipSelect.Option`.
 */
export const ChipSelectChip = forwardRef<HTMLInputElement, ChipSelectChip.Props>(
  (
    {
      'aria-label': ariaLabel,
      children,
      icon,
      isExclusive = false,
      maxWidth,
      onChange,
      onClick,
      overflow,
      readOnly,
      size,
      ...rest
    },
    ref,
  ) => {
    const handleChange = useCallback<ChangeEventHandler<HTMLInputElement>>(
      (event) => {
        // If the chip is read-only, do nothing.
        if (readOnly) {
          event.preventDefault()
          return
        }

        // We only need to deselect other options if the one whose checked state is changing
        // has been checked. If it has been unchecked, there's nothing to do.
        if (event.currentTarget.checked) {
          maybeDeselectOtherOptions(event.currentTarget)
        }
        onChange?.(event)
      },
      [onChange, readOnly],
    )

    const handleClick = useCallback<MouseEventHandler<HTMLLabelElement>>(
      (event) => {
        // Since the read-only attribute has no effect on checkbox inputs, we need to manually prevent
        // click events on chips that have been marked as read-only in order to prevent the preventDefault
        // behaviour of changing their checked state.
        if (readOnly) {
          event.preventDefault()
        }
        onClick?.(event)
      },
      [onClick, readOnly],
    )

    return (
      <ElChipSelectChip
        aria-label={ariaLabel}
        data-size={size}
        onClick={handleClick}
        style={{ maxWidth: maxWidth ? `var(${maxWidth})` : undefined }}
      >
        <ElChipSelectChipInput
          {...rest}
          // NOTE: this attribute is tightly coupled to the `isExclusiveOption` helper.
          data-exclusive={isExclusive}
          onChange={handleChange}
          readOnly={readOnly}
          ref={ref}
          type="checkbox"
        />
        {icon && <ElChipSelectChipIconContainer aria-hidden>{icon}</ElChipSelectChipIconContainer>}
        {children && <ElChipSelectChipLabelText data-overflow={overflow}>{children}</ElChipSelectChipLabelText>}
      </ElChipSelectChip>
    )
  },
)

ChipSelectChip.displayName = 'ChipSelectOption'
