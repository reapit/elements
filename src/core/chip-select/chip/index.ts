export * from './chip'
export * from './determine-next-controlled-state'
export * from './styles'
