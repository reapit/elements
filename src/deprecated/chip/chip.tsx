import { FC, HTMLAttributes, InputHTMLAttributes } from 'react'
import { cx } from '@linaria/core'
import {
  ElDeprecatedChipCheckbox as ElChipCheckbox,
  ElDeprecatedChipLabel as ElChipLabel,
  ElDeprecatedChipGroup as ElChipGroup,
  ElDeprecatedChipGroupInner as ElChipGroupInner,
} from './__styles__'
import { useId } from '../../storybook/random-id'

/** @deprecated */
export interface ChipProps extends InputHTMLAttributes<HTMLInputElement> {}

/** @deprecated */
export const DeprecatedChip: FC<ChipProps> = ({ children, className, id, ...rest }) => {
  const chipId = useId(id)
  return (
    <>
      <ElChipCheckbox id={id ?? chipId} type="checkbox" {...rest} />
      <ElChipLabel htmlFor={id ?? chipId} className={cx(className)}>
        {children}
      </ElChipLabel>
    </>
  )
}

/** @deprecated */
export const DeprecatedChipGroup: FC<HTMLAttributes<HTMLDivElement>> = ({ children, className, ...rest }) => (
  <ElChipGroup className={cx(className)} {...rest}>
    <ElChipGroupInner>{children}</ElChipGroupInner>
  </ElChipGroup>
)
