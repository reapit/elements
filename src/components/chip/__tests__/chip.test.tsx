import { Chip } from '../chip'
import { expect, test, vi } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'

test('renders as a <button> element', () => {
  render(<Chip variant="filter">Label</Chip>)
  expect(screen.getByRole('button', { name: 'Label' })).toBeVisible()
})

test('filter chip has `data-variant="filter"` attribute', () => {
  render(<Chip variant="filter">Label</Chip>)
  expect(screen.getByRole('button', { name: 'Label' })).toHaveAttribute('data-variant', 'filter')
})

test('selection chip has `data-variant="selection"` attribute', () => {
  render(<Chip variant="selection">Label</Chip>)
  expect(screen.getByRole('button', { name: 'Label' })).toHaveAttribute('data-variant', 'selection')
})

test('chip has `type="button"` attribute', () => {
  render(<Chip variant="selection">Label</Chip>)
  expect(screen.getByRole('button', { name: 'Label' })).toHaveAttribute('type', 'button')
})

test('chip has no `aria-disabled` attribute when enabled', () => {
  render(<Chip variant="selection">Label</Chip>)
  expect(screen.getByRole('button', { name: 'Label' })).not.toHaveAttribute('aria-disabled')
})

test('chip label has `data-will-truncate="true"` attribute when `willTruncateLabel` is provided', () => {
  render(
    <Chip willTruncateLabel variant="selection">
      Label
    </Chip>,
  )
  expect(screen.getByText('Label')).toHaveAttribute('data-will-truncate', 'true')
})

test('disabled chip has `aria-disabled="true"` attribute', () => {
  render(
    <Chip isDisabled variant="filter">
      Label
    </Chip>,
  )
  expect(screen.getByRole('button', { name: 'Label' })).toHaveAttribute('aria-disabled', 'true')
})

test('disabled chip does not call `onClick`', () => {
  const fakeClick = vi.fn()
  render(
    <Chip isDisabled onClick={fakeClick} variant="filter">
      Label
    </Chip>,
  )
  const chip = screen.getByRole('button', { name: 'Label' })
  fireEvent.click(chip)

  expect(fakeClick).not.toHaveBeenCalled()
})

test('disabled chip prevents click events from propagating', () => {
  const parentClickHandler = vi.fn()
  render(
    <div onClick={parentClickHandler}>
      <Chip isDisabled variant="filter">
        Label
      </Chip>
    </div>,
  )
  const chip = screen.getByRole('button', { name: 'Label' })
  fireEvent.click(chip)

  expect(parentClickHandler).not.toHaveBeenCalled()
})
