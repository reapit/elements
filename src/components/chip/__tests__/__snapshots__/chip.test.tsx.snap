// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chip component > should match a snapshot 1`] = `
<DocumentFragment>
  <input
    class="mocked-styled-0 el-chip-checkbox"
    id="test-static-id"
    type="checkbox"
  />
  <label
    class="mocked-styled-1 el-chip-label"
    for="test-static-id"
  >
    Some Content
  </label>
</DocumentFragment>
`;

exports[`ChipGroup component > should match a snapshot 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-2 el-chip-group"
  >
    <div
      class="mocked-styled-3 el-chip-group-inner"
    >
      <input
        class="mocked-styled-0 el-chip-checkbox"
        id="test-static-id"
        type="checkbox"
      />
      <label
        class="mocked-styled-1 el-chip-label"
        for="test-static-id"
      >
        Some Content
      </label>
      <input
        class="mocked-styled-0 el-chip-checkbox"
        id="test-static-id"
        type="checkbox"
      />
      <label
        class="mocked-styled-1 el-chip-label"
        for="test-static-id"
      >
        Some Content
      </label>
    </div>
  </div>
</DocumentFragment>
`;
